File: bot.js
const { Client, GatewayIntentBits } = require("discord.js");
const WebSocket = require("ws");
const http = require("http");
const fs = require("fs");
const path = require("path");

// ─── CONFIG ─────────────────────────────────────────────────────────────────
const DISCORD_TOKEN = process.env.DISCORD_TOKEN || "TON_TOKEN_ICI";
const CHANNEL_ID = process.env.CHANNEL_ID || "1507734070535258154";
const WS_PORT = process.env.WS_PORT || 8765;
const HTTP_PORT = process.env.HTTP_PORT || 8766;
// ────────────────────────────────────────────────────────────────────────────

// Serveur HTTP pour servir l'app display
const httpServer = http.createServer((req, res) => {
  const displayPath = path.join(__dirname, "../display/index.html");
  if (req.url === "/" || req.url === "/index.html") {
    fs.readFile(displayPath, (err, data) => {
      if (err) {
        res.writeHead(404);
        res.end("Display app not found. Place index.html in /display/");
        return;
      }
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end(data);
    });
  } else {
    res.writeHead(404);
    res.end("Not found");
  }
});

httpServer.listen(HTTP_PORT, () => {
  console.log(`📺 Display app accessible sur http://localhost:${HTTP_PORT}`);
});

// Serveur WebSocket
const wss = new WebSocket.Server({ server: httpServer });
const clients = new Set();

wss.on("connection", (ws) => {
  clients.add(ws);
  console.log(`✅ Nouveau client connecté (${clients.size} total)`);

  ws.on("close", () => {
    clients.delete(ws);
    console.log(`❌ Client déconnecté (${clients.size} restants)`);
  });
});

function broadcast(data) {
  const payload = JSON.stringify(data);
  clients.forEach((client) => {
    if (client.readyState === WebSocket.OPEN) {
      client.send(payload);
    }
  });
  console.log(`📡 Diffusé à ${clients.size} client(s)`);
}

// Bot Discord
const discord = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
  ],
});

discord.once("ready", () => {
  console.log(`🤖 Bot connecté en tant que ${discord.user.tag}`);
  console.log(`👂 Écoute du salon: ${CHANNEL_ID}`);
});

discord.on("messageCreate", async (message) => {
  // Ignorer les bots et les autres salons
  if (message.bot) return;
  if (message.channel.id !== CHANNEL_ID) return;

  const payload = {
    type: "notification",
    author: message.author.username,
    avatar: message.author.displayAvatarURL({ size: 128 }),
    text: message.content || "",
    images: [],
    timestamp: Date.now(),
  };

  // Récupérer les images attachées
  message.attachments.forEach((attachment) => {
    if (attachment.contentType?.startsWith("image/")) {
      payload.images.push(attachment.url);
    }
  });

  // Récupérer les images dans les embeds
  message.embeds.forEach((embed) => {
    if (embed.image?.url) payload.images.push(embed.image.url);
    if (embed.thumbnail?.url) payload.images.push(embed.thumbnail.url);
  });

  // Ignorer les messages vides (ni texte ni image)
  if (!payload.text && payload.images.length === 0) return;

  console.log(
    `📨 Message de ${payload.author}: "${payload.text}" (${payload.images.length} image(s))`
  );
  broadcast(payload);
});

discord.login(DISCORD_TOKEN).catch((err) => {
  console.error("❌ Impossible de se connecter à Discord:", err.message);
  process.exit(1);
});

File: bot.py
"""
Discord Overlay Bot - Version Stable pour Railway
Lit un salon Discord et diffuse via WebSocket sécurisé.
"""

import discord
import asyncio
import websockets
import json
import os
from datetime import datetime

# ─────────────────────────────────────────────
#  CONFIG — Via variables d'environnement sur Railway
# ─────────────────────────────────────────────
# Sur Railway, configurez ces variables dans l'onglet "Variables"
BOT_TOKEN  = os.environ.get("TOKEN", "VOTRE_TOKEN_LOCAL_SI_TEST")
CHANNEL_ID = int(os.environ.get("CHANNEL_ID", 1507734070535258154))
PORT       = int(os.environ.get("PORT", 8765)) # Railway fournit le port automatiquement
# ─────────────────────────────────────────────

intents = discord.Intents.default()
intents.message_content = True
client = discord.Client(intents=intents)

# Liste des overlays connectés (vos amis)
connected_clients = set()

async def broadcast(payload: dict):
    """Envoie le message à tous les amis connectés en même temps."""
    if not connected_clients:
        return
    data = json.dumps(payload, ensure_ascii=False)
    # asyncio.gather envoie à tout le monde en parallèle
    await asyncio.gather(
        *[ws.send(data) for ws in connected_clients],
        return_exceptions=True,
    )

async def ws_handler(websocket):
    """Gère la connexion de l'overlay d'un ami."""
    connected_clients.add(websocket)
    print(f"[WS] Un ami s'est connecté. Total : {len(connected_clients)}")
    try:
        # Attend que le client se déconnecte, sans bloquer
        await websocket.wait_closed()
    finally:
        connected_clients.discard(websocket)
        print(f"[WS] Un ami s'est déconnecté. Restants : {len(connected_clients)}")

@client.event
async def on_ready():
    ch = client.get_channel(CHANNEL_ID)
    name = ch.name if ch else "introuvable"
    print(f"[Discord] Bot en ligne : {client.user}")
    print(f"[Discord] Surveillance du salon : #{name} ({CHANNEL_ID})")

@client.event
async def on_message(message: discord.Message):
    if message.channel.id != CHANNEL_ID or message.author == client.user:
        return

    attachments = []
    for att in message.attachments:
        mime = att.content_type or ""
        if mime.startswith("image/"):
            kind = "image"
        elif mime.startswith("video/"):
            kind = "video"
        elif mime.startswith("audio/"):
            kind = "audio"
        else:
            kind = "file"
        
        attachments.append({
            "url": att.url,
            "filename": att.filename,
            "kind": kind
        })

    payload = {
        "id": str(message.id),
        "author": message.author.display_name,
        "avatar": str(message.author.display_avatar.url),
        "content": message.content,
        "timestamp": datetime.utcnow().isoformat(),
        "attachments": attachments,
    }

    print(f"[Discord] Nouveau message de {payload['author']}")
    await broadcast(payload)

async def main():
    # ping_interval=20 et ping_timeout=20 empêchent Railway de couper la connexion
    async with websockets.serve(ws_handler, "0.0.0.0", PORT, ping_interval=20, ping_timeout=20):
        print(f"[WS] Serveur WebSocket prêt sur le port {PORT}")
        await client.start(BOT_TOKEN)

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("Arrêt du bot.")